sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,319 @@
+          .num_reps = 4,
+      },
+      {
+          .label = "memset",
+          .setup_buf1 = &fill_buf_zeroes,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memset,
+          .num_reps = 4,
+      },
+      // TODO Add test cases that for memcmp and memrcmp that aren't just
+      // zeroes.
+      {
+          .label = "memcmp",
+          .setup_buf1 = &fill_buf_zeroes,
+          .setup_buf2 = &fill_buf_zeroes,
+          .func = &test_memcmp,
+          .num_reps = 4,
+      },
+      {
+          .label = "memrcmp",
+          .setup_buf1 = &fill_buf_zeroes,
+          .setup_buf2 = &fill_buf_zeroes,
+          .func = &test_memrcmp,
+          .num_reps = 4,
+      },
+      // TODO Add test cases for memchr and memrchr that will exercise them
+      // better. Currently, they search for a byte that has equal probability of
+      // showing up anywhere in the buffer. We should also have a test on
+      // pathological inputs where they are forced to search the whole buffer.
+      {
+          .label = "memchr",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memchr,
+          .num_reps = 4,
+      },
+      {
+          .label = "memrchr",
+          .setup_buf1 = &fill_buf_arbitrary_values,
+          .setup_buf2 = &fill_buf_arbitrary_values,
+          .func = &test_memrchr,
+          .num_reps = 4,
+      },
+  };
+
+  size_t max_label_len = 0;
+  for (size_t i = 0; i < ARRAYSIZE(kPerfTests); ++i) {
+    const size_t len = strlen(kPerfTests[i].label);
+    if (len > max_label_len) {
+      max_label_len = len;
+    }
+  }
+
+  test_context_t context;
+  test_context_init(&context);
+
+  for (size_t i = 0; i < ARRAYSIZE(kPerfTests); ++i) {
+    const perf_test_t *test = &kPerfTests[i];
+
+    CHECK(test->num_reps <= PERFTEST_MAX_NUM_REPS);
+    uint64_t durations[PERFTEST_MAX_NUM_REPS] = {0};
+
+    for (size_t rep = 0; rep < test->num_reps; ++rep) {
+      durations[rep] = perf_test_run(&context, test);
+    }
+
+    // Using uint32_t because base_printf() cannot print uint64_t values.
+    uint64_t duration_mean, duration_variance;
+    CHECK(compute_statistics(durations, test->num_reps, &duration_mean,
+                             &duration_variance));
+
+    // Use base_printf() instead of LOG_INFO() because line numbers can
+    // break column alignment.
+    if (duration_mean > UINT32_MAX || duration_variance > UINT32_MAX) {
+      LOG_WARNING(
+          "Cannot print statistics for %s because mean and/or variance do not "
+          "fit in 32 bits",
+          test->label);
+    } else {
+      print_padded(test->label, max_label_len);
+      CHECK(test->num_reps <= UINT32_MAX);
+      base_printf("  %20d cycles (mean) %20d cycles (variance) %20d (n)\n",
+                  (uint32_t)duration_mean, (uint32_t)duration_variance,
+                  (uint32_t)test->num_reps);
+    }
+  }
+  return true;
```

Comment:
I haven't uploaded a PR yet, but I'm working on a correctness-focused set of unit tests. I figured this would be more focused on measuring the performance.

I suppose I could encode expectations in `perf_test_t`. Do you think that's worth doing?

Reply:
I think so. We would like this test to fail if for some reason we revert back to byte-wise implementations, right?

Reply:
Oh! I thought you were talking about checking input/output expectations. Are you talking about checking runtime expectations? Or both?

Reply:
Bizarre... I tried encoding the expected number of cycles and setting `num_reps=1`. I set the expected cycle count to 9020065, but when I recompile and run, the memcpy test took 10019061 cycles. OK, I updated the expectation, but now it took 10515565 cycles. But that's apparently a fixed point, now it consistently takes 10515565 cycles.

Between recompiling, I can rerun the memcpy test any number of times and it's always has the same cycle count as the last run. It seems like editing the source and rerunning the test somehow influences the runtime. Maybe pushing the bitstream to the FPGA re-seeds the RNG or something?

Reply:
I replaced the definition of `fill_buf_arbitrary_values()` with something deterministic and now the cycle count is also deterministic across recompiles. Huh.

Reply:
The SRAM is a 32-bit memory plus ECC bits.  Writing to it in 8-bit or 16-bit chunks causes a read-modify-write behind the scenes, along with recalculating the ECC bits.  There is also scrambling of the RAM so that an attacker with access to the die can't just read out RAM to get the values.

Which RNG were you using to generate random values?  `rand_testutils_gen32` should be deterministic in its outputs, _however_, from run-to-run, its timing may vary because of the internal construction of that routine.

Reply:
`rand_testutils_gen32()` is what I was using. I didn't squash commits yet, so you can still see that version here: https://github.com/lowRISC/opentitan/pull/13871/commits/9b74b9b07eb029030f84daa8601122f39b67a2ef

Aha, so it sounds like scrambling was taking an unpredictable number of cycles? That would certainly explain why hardcoding the cycle-count expectations was not working.

Reply:
I don't think it was the scrambling or ECC calcuation (that was my first thought, but after thinking about it for a bit, I decided that it can't possibly be the hardware causing the cycle variance).

It's `rand_testutils_gen32` that was causing the variance.  That routine isn't reseeded between runs, and the main bit of that routine that creates the random output is:
```
if (lsb) lfsr ^= kConst
```
Since it isn't reseeded between runs, the stream of outputs will be different from run to run and thus the number of times the  internal "mixing" branch is taken will be different from run to run.

This probably could be solved by changing that `if` statement into something like:
```
lfsr ^= lsb ? kConst : 0;
```
But even then, we'd have to look at how the compiler generated code for that conditional.  We'd probably have to use the constant time routines from `hardened.h`:
```
lfsr ^= ct_cmov32(ct_seqz32(lsb), 0, kConst)
```